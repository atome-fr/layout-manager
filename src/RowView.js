import React, {Component} from 'react';

// Import component
import View from './View';

/**
 * Class used to manage a row view
 * @class
 */
class RowView extends Component {

    /**
     * Method for render the components
     * @returns {*}
     */
    render() {
        return (
            <div className="row">
                <View>
                    {this.props.view.component}
                </View>
            </div>
        );
    }
}

export default RowView;