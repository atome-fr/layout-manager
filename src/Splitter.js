import React, {Component} from 'react';

/**
 * Class used to manage a splitter
 * @class
 */
class Splitter extends Component {

    /**
     * Override constructor
     * @param props
     */
    constructor(props) {
        super(props);

        this.handleMouseOrTouchStart = this.handleMouseOrTouchStart.bind(this);
        this.handleMouseOrTouchMove = this.handleMouseOrTouchMove.bind(this);
        this.handleMouseOrTouchEnd = this.handleMouseOrTouchEnd.bind(this);
    }

    /**
     * Handler when the user start to move the splitter
     */
    handleMouseOrTouchStart(ev) {
        if ((ev.clientX && ev.clientY) || (ev.touches && ev.touches.length > 0)) {
            this.mousePos = {x: ev.clientX || ev.touches[0].clientX, y: ev.clientY || ev.touches[0].clientY};
        }
        document.addEventListener('mousemove', this.handleMouseOrTouchMove);
        document.addEventListener('touchmove', this.handleMouseOrTouchMove);
        document.addEventListener('mouseup', this.handleMouseOrTouchEnd);
        document.addEventListener('touchend', this.handleMouseOrTouchEnd);
    }

    /**
     * Handler when the user move the splitter
     * @param ev
     */
    handleMouseOrTouchMove(ev) {
        ev.preventDefault();

        let newMousePos = undefined;
        if ((ev.clientX && ev.clientY) || (ev.touches && ev.touches.length > 0)) {
            newMousePos = {x: ev.clientX || ev.touches[0].clientX, y: ev.clientY || ev.touches[0].clientY};
        }

        if(newMousePos) {
            const offset = {x: newMousePos.x - this.mousePos.x, y: newMousePos.y - this.mousePos.y};
            this.mousePos = newMousePos;
            this.props.onChange(offset, this.el);
        }
    }

    /**
     * Handler when the user end to move the splitter
     */
    handleMouseOrTouchEnd() {
        document.removeEventListener('mousemove', this.handleMouseOrTouchMove);
        document.removeEventListener('touchmove', this.handleMouseOrTouchMove);
        document.removeEventListener('mouseup', this.handleMouseOrTouchEnd);
        document.removeEventListener('touchend', this.handleMouseOrTouchEnd);
    }

    /**
     * Method for render the components
     * @returns {*}
     */
    render() {
        return (
            <div ref={(el) => this.el = el}
                 className={"splitter " + this.props.type}
                 onTouchStart={this.handleMouseOrTouchStart}
                 onMouseDown={this.handleMouseOrTouchStart}>
                <div className="handle">
                </div>
            </div>
        );
    }
}

export default Splitter;