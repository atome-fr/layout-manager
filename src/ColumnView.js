import React, {Component} from 'react';

// Import components
import Container from './Container';

/**
 * Class used to manage a column view and render a container of rows
 * @class
 */
class ColumnView extends Component {

    /**
     * Method for render the components
     * @returns {*}
     */
    render() {
        return (
            <div className="column">

                <Container type="row" views={this.props.views}/>

            </div>
        );
    }
}

export default ColumnView;