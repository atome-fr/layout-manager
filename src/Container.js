import React, {Component} from 'react';
import ReactDOM from 'react-dom';

// Import components
import ColumnView from './ColumnView';
import RowView from './RowView';
import Splitter from './Splitter';

/**
 * Class used to manage a container of components (columns or rows)
 * @class
 */
class Container extends Component {

    /**
     * Override the constructor
     * @param props
     */
    constructor(props) {
        super(props);

        this.handleSplitterChange = this.handleSplitterChange.bind(this);

        this.sliderSize = 10;

        this.splitters = [];
        this.children = [];

        if (props.type === 'column') {
            this.sizePropName = "width";
            this.offsetPropName = "x";
            this.splitterType = "vertical";
        }
        else if (props.type === 'row') {
            this.sizePropName = "height";
            this.offsetPropName = "y";
            this.splitterType = "horizontal";
        }
    }

    /**
     * Component lifecycle method
     */
    componentDidMount() {
        this._computeSize();
        window.addEventListener("resize", () => this._computeSize());
    }

    /**
     * Component lifecycle method
     */
    componentDidUpdate() {
        if (this.children.length > 0) {
            this._computeSize();
        }
    }

    /**
     * Component lifecycle method
     */
    componentWillUnmount() {
        window.removeEventListener("resize", () => this._computeSize());
    }

    /**
     * Handler when the splitter is moved
     * @param splitterOffset The position of the slider
     * @param splitterNode The splitter node
     */
    handleSplitterChange(splitterOffset, splitterNode) {
        const sizeAvailable = this.container[`client${this.sizePropName.charAt(0).toUpperCase() + this.sizePropName.slice(1)}`] - (this.splitters.length * this.sliderSize);
        const ratioOffset = splitterOffset[this.offsetPropName] / sizeAvailable;
        const splitterIndex = this._searchSplitterIndex(splitterNode);

        this._updateRatios(splitterIndex, ratioOffset);

        this._applyRatios(sizeAvailable);
    }

    /**
     * Method for update the ratio for the view components
     * @param index The index of the ratio to change
     * @param ratio The new ratio to change
     * @private
     */
    _updateRatios(index, ratio) {
        const newRatios = this.ratios.slice();

        let ratioToAdd = ratio;
        let indexToAdd = index;
        while (ratioToAdd !== 0 && indexToAdd >= 0) {
            if (newRatios[indexToAdd] + ratioToAdd > 0) {
                newRatios[indexToAdd] += ratioToAdd;
                ratioToAdd = 0;
            } else {
                ratioToAdd += newRatios[indexToAdd];
                newRatios[indexToAdd] = 0;
                indexToAdd--;
            }
        }

        let ratioToRem = ratio;
        let indexToRem = index + 1;
        while (ratioToRem !== 0 && indexToRem < this.children.length) {
            if (newRatios[indexToRem] - ratioToRem > 0) {
                newRatios[indexToRem] -= ratioToRem;
                ratioToRem = 0;
            } else {
                ratioToRem -= newRatios[indexToRem];
                newRatios[indexToRem] = 0;
                indexToRem++;
            }
        }

        const sumRatios = newRatios.reduce((acc, val) => acc + val, 0);
        if (sumRatios < 0.99999 || sumRatios > 1.00001) {
            return;
        }

        this.ratios = newRatios;
    }

    /**
     * Method for apply the ratio to the components
     * @param sizeAvailable The available global size for apply each ratio
     * @private
     */
    _applyRatios(sizeAvailable) {
        for (let i = 0; i < this.children.length; i++) {
            ReactDOM.findDOMNode(this.children[i]).style[this.sizePropName] = sizeAvailable * this.ratios[i] + 'px';
        }
    }

    /**
     * Method for search the index of the splitter in the array of refs
     * @param splitterNode The node of the splitter moved
     * @returns {number} The index of the splitter
     * @private
     */
    _searchSplitterIndex(splitterNode) {
        for (let i = 0; i < this.splitters.length; i++) {
            if (ReactDOM.findDOMNode(this.splitters[i]) === splitterNode) {
                return i;
            }
        }
    }

    /**
     * Method for compute the size for each column
     * @private
     */
    _computeSize() {
        this.ratios = [];

        const sizeAvailable = this.container[`client${this.sizePropName.charAt(0).toUpperCase() + this.sizePropName.slice(1)}`] - (this.splitters.length * this.sliderSize);

        for (let i = 0; i < this.children.length - 1; i++) {
            this.ratios[i] = (sizeAvailable / this.children.length) / sizeAvailable;
        }
        // In order to have exactly a ratio of one
        this.ratios[this.children.length - 1] = 1 - this.ratios.reduce((acc, val) => {
            return acc + val;
        }, 0);

        this._applyRatios(sizeAvailable);
    }

    /**
     * Method for generate the render with the right numbers of columns and splitters
     * @returns {Array}
     * @private
     */
    _generateRender() {
        let toRender = [];

        // Reset array of refs
        this.splitters = [];
        this.children = [];

        let index = 0;
        this.props.views.forEach((view) => {
            if (view.visible) {
                if (index > 0) {
                    toRender.push(<Splitter ref={(splitter) => {
                        if (splitter) {
                            this.splitters.push(splitter);
                        }
                    }}
                                            key={'splitter_' + this.props.type + '_' + index}
                                            type={this.splitterType}
                                            onChange={(offset, node) => this.handleSplitterChange(offset, node)}/>);
                }

                if (this.props.type === "column") {
                    toRender.push(<ColumnView ref={(col) => {
                        if (col) {
                            this.children.push(col);
                        }
                    }}
                                              key={this.props.type + '_' + index}
                                              views={view.components}/>);
                }
                else if (this.props.type === "row") {
                    toRender.push(<RowView ref={(row) => {
                        if (row) {
                            this.children.push(row);
                        }
                    }}
                                           key={'row_' + index}
                                           view={view}/>);
                }

                index++;
            }
        });

        return toRender;
    }

    /**
     * Method for render the components
     * @returns {*}
     */
    render() {
        const classNames = (this.props.type === "column" ? "container parent" : "container");

        return (
            <div ref={(el) => this.container = el}
                 className={classNames}>

                {this._generateRender()}

            </div>
        );
    }
}

export default Container;