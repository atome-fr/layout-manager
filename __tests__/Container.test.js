import React from 'react';
import ReactDOM from 'react-dom';
import Container from '../src/Container';
import ColumnView from '../src/ColumnView';
import RowView from '../src/RowView';
import Splitter from '../src/Splitter';
import {shallow} from 'enzyme';
import toJson from 'enzyme-to-json';

describe('Container Component', () => {
    let oldVal = null;

    beforeEach(() => {
        oldVal = Container.prototype.componentDidMount;
        Container.prototype.componentDidMount = jest.fn();
    });

    afterEach(() => {
        Container.prototype.componentDidMount = oldVal;
    });

    describe('UI Render Test', () => {

        it('should render with ColumnView children component', () => {
            const views = [
                {
                    visible: true,
                    components: [
                        {
                            name: 'rendering',
                            visible: true
                        },
                        {
                            name: 'ide',
                            visible: true
                        }
                    ]
                },
                {
                    visible: true,
                    components: [
                        {
                            name: 'nodal',
                            visible: true
                        }
                    ]
                }
            ];

            const wrapper = shallow(<Container type="column" views={views}/>);
            const tree = toJson(wrapper);

            expect(tree).toMatchSnapshot();
            expect(wrapper.find(ColumnView).length).toBe(2);
            expect(wrapper.find(Splitter).length).toBe(1);
        });

        it('should render with RowView children component', () => {
            const views = [
                {
                    name: 'rendering',
                    visible: true
                },
                {
                    name: 'ide',
                    visible: true
                }
            ];

            const wrapper = shallow(<Container type="row" views={views}/>);
            const tree = toJson(wrapper);

            expect(tree).toMatchSnapshot();
            expect(wrapper.find(RowView).length).toBe(2);
            expect(wrapper.find(Splitter).length).toBe(1);
        });
    });

    describe('Component Method Test', () => {

        it('should call the handleSplitterChange method and call the methods for update the size of the different components', () => {
            const oldVal2 = Container.prototype._searchSplitterIndex;
            Container.prototype._searchSplitterIndex = jest.fn(() => 0);

            const oldVal3 = Container.prototype._updateRatios;
            Container.prototype._updateRatios = jest.fn();

            const oldVal4 = Container.prototype._applyRatios;
            Container.prototype._applyRatios = jest.fn();

            const wrapper = shallow(<Container type="column" views={[]}/>);

            wrapper.instance().container = {
                clientWidth: 300
            };

            wrapper.instance().handleSplitterChange({x: 1, y: 1}, {});

            expect(Container.prototype._searchSplitterIndex).toHaveBeenCalledTimes(1);
            expect(Container.prototype._searchSplitterIndex).toHaveBeenCalledWith({});
            expect(Container.prototype._updateRatios).toHaveBeenCalledTimes(1);
            expect(Container.prototype._updateRatios).toHaveBeenCalledWith(0, (1 / 300));
            expect(Container.prototype._applyRatios).toHaveBeenCalledTimes(1);
            expect(Container.prototype._applyRatios).toHaveBeenCalledWith(300);

            Container.prototype._searchSplitterIndex = oldVal2;
            Container.prototype._updateRatios = oldVal3;
            Container.prototype._applyRatios = oldVal4;
        });

        it('should call the _updateRatios method and update the ratio of the container', () => {
            const wrapper = shallow(<Container type="column" views={[]}/>);

            wrapper.instance().ratios = [0.2, 0.2, 0.6];
            wrapper.instance().children = {
                length: 3
            };

            // Should update the ratios
            wrapper.instance()._updateRatios(0, 0.2);
            expect(wrapper.instance().ratios).toEqual([0.4, 0, 0.6]);

            // Shouldn't update the ratios
            wrapper.instance()._updateRatios(0, 2);
            expect(wrapper.instance().ratios).toEqual([0.4, 0, 0.6]);
        });

        it('should call the _updateRatios method and update the ratio of the container', () => {
            const wrapper = shallow(<Container type="column" views={[]}/>);

            wrapper.instance().ratios = [0.2, 0.2, 0.6];
            wrapper.instance().children = {
                length: 3
            };

            // Should update the ratios
            wrapper.instance()._updateRatios(0, 0.2);
            expect(wrapper.instance().ratios).toEqual([0.4, 0, 0.6]);

            // Shouldn't update the ratios
            wrapper.instance()._updateRatios(0, 2);
            expect(wrapper.instance().ratios).toEqual([0.4, 0, 0.6]);
        });

        it('should call the _applyRatios method and set the correct size to the children', () => {
            const child1 = {
                style: {
                    width: 0
                }
            };

            const child2 = {
                style: {
                    width: 0
                }
            };

            const oldVal = ReactDOM.findDOMNode;
            ReactDOM.findDOMNode = jest.fn().mockReturnValueOnce(child1).mockReturnValueOnce(child2);

            const wrapper = shallow(<Container type="column" views={[]}/>);

            wrapper.instance().ratios = [0.5, 0.5];
            wrapper.instance().children = [child1, child2];

            wrapper.instance()._applyRatios(300);

            expect(child1.style.width).toBe('150px');
            expect(child2.style.width).toBe('150px');

            ReactDOM.findDOMNode = oldVal;
        });

        it('should call the _searchSplitterIndex method and return the index of the splitter', () => {
            const splitter1 = "splitter1";
            const splitter2 = "splitter2";

            const oldVal = ReactDOM.findDOMNode;
            ReactDOM.findDOMNode = jest.fn().mockReturnValueOnce(splitter1).mockReturnValueOnce(splitter1).mockReturnValueOnce(splitter2);

            const wrapper = shallow(<Container type="column" views={[]}/>);

            wrapper.instance().splitters = [splitter1, splitter2];

            expect(wrapper.instance()._searchSplitterIndex("splitter1")).toBe(0);
            expect(wrapper.instance()._searchSplitterIndex("splitter2")).toBe(1);

            ReactDOM.findDOMNode = oldVal;
        });

        it('should call the _computeSize method and set the ratios for the components of the container', () => {
            const oldVal2 = Container.prototype._applyRatios;
            Container.prototype._applyRatios = jest.fn();

            const oldVal3 = document.dispatchEvent;
            document.dispatchEvent = jest.fn();

            const wrapper = shallow(<Container type="column" views={[]}/>);

            wrapper.instance().container = {
                clientWidth: 300
            };
            wrapper.instance().children = {
                length: 4
            };

            wrapper.instance()._computeSize(true);

            expect(wrapper.instance().ratios).toEqual([0.25, 0.25, 0.25, 0.25]);
            expect(Container.prototype._applyRatios).toHaveBeenCalledTimes(1);
            expect(document.dispatchEvent).toHaveBeenCalledTimes(0);
            expect(Container.prototype._applyRatios).toHaveBeenCalledWith(300);

            wrapper.instance()._computeSize();
            expect(document.dispatchEvent).toHaveBeenCalledTimes(1);

            Container.prototype._applyRatios = oldVal2;
            document.dispatchEvent = oldVal3;
        });

        it('should call the _generateRender method and return the correct array of ColumnView components to render', () => {
            const views = [
                {
                    visible: true,
                    components: [
                        {
                            name: 'rendering',
                            visible: true
                        },
                        {
                            name: 'ide',
                            visible: true
                        }
                    ]
                },
                {
                    visible: true,
                    components: [
                        {
                            name: 'nodal',
                            visible: true
                        }
                    ]
                }
            ];

            const wrapper = shallow(<Container type="column" views={views}/>);

            const toRender = wrapper.instance()._generateRender();

            expect(toRender.length).toBe(3);
            expect(toRender[0].type).toBe(ColumnView);
            expect(toRender[1].type).toBe(Splitter);
            expect(toRender[2].type).toBe(ColumnView);
        });

        it('should call the _generateRender method and return the correct array of RowView components to render', () => {
            const views = [
                {
                    name: 'rendering',
                    visible: true
                },
                {
                    name: 'ide',
                    visible: true
                }
            ];

            const wrapper = shallow(<Container type="row" views={views}/>);

            const toRender = wrapper.instance()._generateRender();

            expect(toRender.length).toBe(3);
            expect(toRender[0].type).toBe(RowView);
            expect(toRender[1].type).toBe(Splitter);
            expect(toRender[2].type).toBe(RowView);
        });

    });

});