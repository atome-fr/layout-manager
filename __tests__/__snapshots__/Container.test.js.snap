// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Container Component UI Render Test should render with ColumnView children component 1`] = `
<div
  className="container parent"
>
  <ColumnView
    key="column_0"
    views={
      Array [
        Object {
          "name": "rendering",
          "visible": true,
        },
        Object {
          "name": "ide",
          "visible": true,
        },
      ]
    }
  />
  <Splitter
    key="splitter_column_1"
    onChange={[Function]}
    type="vertical"
  />
  <ColumnView
    key="column_1"
    views={
      Array [
        Object {
          "name": "nodal",
          "visible": true,
        },
      ]
    }
  />
</div>
`;

exports[`Container Component UI Render Test should render with RowView children component 1`] = `
<div
  className="container"
>
  <RowView
    key="row_0"
    view={
      Object {
        "name": "rendering",
        "visible": true,
      }
    }
  />
  <Splitter
    key="splitter_row_1"
    onChange={[Function]}
    type="horizontal"
  />
  <RowView
    key="row_1"
    view={
      Object {
        "name": "ide",
        "visible": true,
      }
    }
  />
</div>
`;
